auth/src/main/java/com/firebase/ui/auth/util/smartlock/SignInDelegate.java

```diff
@@ -0,0 +1,427 @@
+            }
+        } else {
+            startActivityForResult(
+                    AuthMethodPickerActivity.createIntent(
+                            getContext(),
+                            mHelper.getFlowParams()),
+                    RC_AUTH_METHOD_PICKER);
+        }
+    }
+
+    /**
+     * Begin sign in process with email and password from a SmartLock credential.
+     * On success, finish with {@code RESULT_OK}.
+     * On failure, delete the credential from SmartLock (if applicable) and then launch the
+     * auth method picker flow.
+     */
+    private void signInWithEmailAndPassword(String email, String password) {
+        mHelper.getFirebaseAuth()
+                .signInWithEmailAndPassword(email, password)
+                .addOnFailureListener(new TaskFailureLogger(
+                        TAG, "Error signing in with email and password"))
+                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
+                    @Override
+                    public void onSuccess(AuthResult authResult) {
+                        finish(RESULT_OK, new Intent());
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        if (e instanceof FirebaseAuthInvalidUserException) {
+                            // In this case the credential saved in SmartLock was not
+                            // a valid credential, we should delete it from SmartLock
+                            // before continuing.
+                            deleteCredentialAndRedirect();
+                        } else {
+                            startAuthMethodChoice();
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Delete the last credential retrieved from SmartLock and then redirect to the
+     * auth method choice flow.
+     */
+    private void deleteCredentialAndRedirect() {
+        if (mCredential == null) {
+            Log.w(TAG, "deleteCredentialAndRedirect: null credential");
+            startAuthMethodChoice();
+            return;
+        }
+
+        CredentialsApiHelper credentialsApiHelper = CredentialsApiHelper.getInstance(getActivity());
+        credentialsApiHelper.delete(mCredential)
+                .addOnCompleteListener(new OnCompleteListener<Status>() {
+                    @Override
+                    public void onComplete(@NonNull Task<Status> task) {
+                        if (!task.isSuccessful()) {
+                            Log.w(TAG, "deleteCredential:failure", task.getException());
+                        }
+                        startAuthMethodChoice();
+                    }
+                });
+    }
+
+    private void redirectToIdpSignIn(String email, String accountType) {
+        Intent nextIntent;
+        switch (accountType) {
```

Comment:
nit: don't we have a "provider translate" helper funciton somewhere?  In which case all the non-default cases of this switch could be:
   `nextIntent = IdpSignInContainerActivity.createIntent(getContext(), mHelper.getFlowParams(), translateProvider(accountType), email)`